File: LeetCode-July-Challenge/ValidAnagram.js
var isAnagram = function(s, t) {
    if(s.length !== t.length) return false;
    const count = {};
    const n = s.length;
    for(let i = 0; i < n; ++i){
        if(!count[t[i]]) count[t[i]] = 0;
        if(!count[s[i]]) count[s[i]] = 0;
        count[t[i]]++;
        count[s[i]]--;
    }
    for(let ch in count){
        if(count[ch] !== 0) return false;
    }
    return true;
};
let s = "rat", t = "car";
let s1 = "anagram", t1 = "nagaram"
console.log(isAnagram(s, t));
console.log(isAnagram(s1, t1));

File: LeetCode-July-Challenge/ValidAnagram.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool isAnagram(string s, string t) {
        int lenS = s.size();
        int lenT = t.size();
        if(lenS != lenT) return false;
        vector<int> tFreq(26, 0);
        vector<int> window(26, 0);
        for(int i = 0; i < lenT; ++i){
            tFreq[t[i] - 'a']++;
            window[s[i] - 'a']++;
        }
        if(tFreq != window) return false;
        for(int i = lenT; i < lenS; ++i){
            window[s[i - lenT] - 'a']--;
            window[s[i] - 'a']++;
            if(tFreq != window) return false;
        }
        return true;
    }
};
int main(){
    string s = "anagram", t = "nagaram";
    Solution S;
    cout<<std::boolalpha<<S.isAnagram(s, t)<<endl;
    return 0;
}